_drafts/gnu-build-id-for-firmware.md

```diff
@@ -0,0 +1,270 @@
+  Owner                 Data size       Description
+  GNU                  0x00000014       NT_GNU_BUILD_ID (unique build ID
+bitstring)
+    Build ID: bab6b09f86b3c3017499d8e386447a610c559bd5
+```
+
+As you can see, our binary now contains the build id
+`bab6b09f86b3c3017499d8e386447a610c559bd5`.
+
+## Bundling the GNU build ID in firmware bin files
+
+Getting the build ID in your executables on Linux is easy: they are ELF files!
+Firmware on the other hand typically deals with binaries which are assembling by
+copying relevant sections of the elf at the right offset in a file.
+
+This is typically accomplished with objcopy:
+```terminal
+$ arm-none-eabi-objcopy firmware.elf firmware.bin -O binary
+```
+
+This takes every elf section earmarked to be loaded and places them at 
+the correct offset in the bin file. In the process, most debug sections are stripped out.
+
+Dumping the elf sections of the resulting `minimal.elf` gives us:
+
+```
+build/minimal.elf:     file format elf32-littlearm
+
+Sections:
+Idx Name          Size      VMA       LMA       File off  Algn
+  0 .note.gnu.build-id 00000024  00000000  00000000  00010000  2**2
+                  CONTENTS, ALLOC, LOAD, READONLY, DATA
+  1 .text         000004e4  00000024  00000024  00010024  2**2
+                  CONTENTS, ALLOC, LOAD, READONLY, CODE
+  2 .bss          00000000  20000000  20000000  00000000  2**0
+                  ALLOC
+  3 .data         00000000  20000000  20000000  00010508  2**0
+                  CONTENTS, ALLOC, LOAD, DATA
+  4 .stack        00002000  20000000  20000000  00020000  2**0
+                  ALLOC
+  5 .debug_info   00004076  00000000  00000000  00010508  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  6 .debug_abbrev 00000ac3  00000000  00000000  0001457e  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  7 .debug_aranges 000000f8  00000000  00000000  00015041  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  8 .debug_ranges 000000b8  00000000  00000000  00015139  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  9 .debug_line   00000cfa  00000000  00000000  000151f1  2**0
+                  CONTENTS, READONLY, DEBUGGING
+ 10 .debug_str    0000111a  00000000  00000000  00015eeb  2**0
+                  CONTENTS, READONLY, DEBUGGING
+ 11 .comment      00000075  00000000  00000000  00017005  2**0
+                  CONTENTS, READONLY
+ 12 .ARM.attributes 0000002c  00000000  00000000  0001707a  2**0
+                  CONTENTS, READONLY
+ 13 .debug_frame  00000298  00000000  00000000  000170a8  2**2
+                  CONTENTS, READONLY, DEBUGGING
+```
+
+As you can see, the `.text`, `.bss`, `.data`, `.stack` sections each have the 
+`LOAD` attribute, all others (including our `.note.gnu.build-id`) do not and will be
+discarded.
+
+To add a section to our binary, we must specify an address for it in our linker
+script. Assuming your linker script declares the following memory layout:
+
+```
+MEMORY¬
```

Comment:
I'd probably drop the newline markers `¬` so someone can just copy/paste

Reply:
Fixed